src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -1204,19 +1240,214 @@ private async Task PingServer()
+
+                if (!TryChangeState(HubConnectionState.Connected, HubConnectionState.Disconnected))
+                {
+                    return;
+                }
+
+                RunCloseEvent(connectionState.CloseException);
+            }
+            else
+            {
+                // Fire-and-forget the reconnect loop
+                _ = Reconnect(connectionState.CloseException);
+            }
+        }
+
+        private void RunCloseEvent(Exception closeException)
+        {
+            var closed = Closed;
+
+            async Task RunClosedEventAsync()
+            {
+                // Dispatch to the thread pool before we invoke the user callback
+                await AwaitableThreadPool.Yield();
+
+                try
+                {
+                    Log.InvokingClosedEventHandler(_logger);
+                    await closed.Invoke(closeException);
+                }
+                catch (Exception ex)
+                {
+                    Log.ErrorDuringClosedEvent(_logger, ex);
+                }
+            }
+
+            // There is no need to start a new task if there is no Closed event registered
+            if (closed != null)
+            {
+                // Fire-and-forget the closed event
+                _ = RunClosedEventAsync();
+            }
+        }
+
+        private async Task Reconnect(Exception closeException)
+        {
+            var reconnectStartTime = DateTime.UtcNow;
+            var previousReconnectAttempts = 0;
+
+            var nextRetryDelay = GetNextRetryDelay(previousReconnectAttempts++, TimeSpan.Zero);
+
+            if (nextRetryDelay == null) {
+                //_logger.log(LogLevel.Debug, "Connection not reconnecting because the IReconnectPolicy returned null on the first reconnect attempt.");
+                RunCloseEvent(closeException);
+                return;
+            }
+
+            if (!TryChangeState(HubConnectionState.Connected, HubConnectionState.Reconnecting))
+            {
+                return;
+            }
+
+            if (closeException != null) {
+                //_logger.log(LogLevel.Information, `Connection reconnecting because of error '${error}'.`);
+            } else {
+                //_logger.log(LogLevel.Information, "Connection reconnecting.");
+            }
+
+            RunReconnectingEvent(closeException);
+
+            while (nextRetryDelay != null) {
+                // _logger.log(LogLevel.Information, `Reconnect attempt number ${previousReconnectAttempts} will start in ${nextRetryDelay} ms.`);
+
+                await Task.Delay(nextRetryDelay.Value);
+
+                if (State != HubConnectionState.Reconnecting) {
+                    //_logger.log(LogLevel.Debug, "Connection left the reconnecting state during reconnect delay. Done reconnecting.");
+                    return;
+                }
+
+                try {
+                    // TODO: Use a token that is canceled in StopAsync.
```

Comment:
I still need to do work to ensure stop always works while the HubConnection is reconnecting. And I also need to add tests of course.